File: src/lib/types/response.ts
/** Structured AI response contract (mirrors `bluey_core::types::response`). */

export type ResponseType = "answer" | "code" | "system-design" | "summary" | "suggestion" | "research";

export interface ResponseSection {
  id: string;
  title: string;
  content: string;
  /** Rendering hint. */
  kind?: "text" | "list" | "code" | "diagram" | "table" | "calculation";
  language?: string;
  collapsed?: boolean;
}

export interface Citation {
  id: string;
  title: string;
  url: string;
  snippet?: string;
}

export interface CodeBlock {
  language: string;
  code: string;
  filename?: string;
}

export type FeedbackRating = "up" | "down";
export type FeedbackCategory = "wrong" | "too_long" | "not_relevant" | "missed_context" | "wrong_tone";

export interface ResponseFeedback {
  responseId: string;
  rating: FeedbackRating;
  categories?: FeedbackCategory[];
  comment?: string;
  createdAt: string;
}

export interface ResponseMetrics {
  provider?: string;
  model?: string;
  captureMs?: number;
  ocrMs?: number;
  accessibilityMs?: number;
  contextAssemblyMs?: number;
  timeToFirstTokenMs?: number;
  totalMs?: number;
  inputTokens?: number;
  outputTokens?: number;
  contextTokens?: number;
}

export interface BlueyResponse {
  id: string;
  requestId: string;
  sessionId?: string;
  modeId: string;
  type: ResponseType;
  title?: string;
  /** Markdown body. Code blocks inside are preserved verbatim. */
  content: string;
  code?: CodeBlock;
  sections?: ResponseSection[];
  citations?: Citation[];
  confidence?: number;
  /** The user instruction or detected question this responds to. */
  prompt?: string;
  /** Mermaid diagram source when a system-design response includes one. */
  diagram?: string;
  metrics?: ResponseMetrics;
  feedback?: ResponseFeedback;
  /** True when this response was prepared proactively and not yet shown. */
  prepared?: boolean;
  createdAt: string;
}

/** The raw structured object we ask models to return (provider structured output). */
export interface StructuredModelOutput {
  responseType: ResponseType;
  title?: string;
  content: string;
  sections?: Array<Omit<ResponseSection, "id">>;
  code?: CodeBlock;
  diagram?: string;
  confidence?: number;
  citations?: Array<Omit<Citation, "id">>;
}

File: src-tauri/crates/bluey-core/src/types/response.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize, Default)]
#[serde(rename_all = "kebab-case")]
pub enum ResponseType {
    #[default]
    Answer,
    Code,
    SystemDesign,
    Summary,
    Suggestion,
    Research,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SectionKind {
    Text,
    List,
    Code,
    Diagram,
    Table,
    Calculation,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResponseSection {
    pub id: String,
    pub title: String,
    pub content: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub kind: Option<SectionKind>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub language: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub collapsed: Option<bool>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Citation {
    pub id: String,
    pub title: String,
    pub url: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub snippet: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CodeBlock {
    pub language: String,
    pub code: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub filename: Option<String>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum FeedbackRating {
    Up,
    Down,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum FeedbackCategory {
    Wrong,
    TooLong,
    NotRelevant,
    MissedContext,
    WrongTone,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResponseFeedback {
    pub response_id: String,
    pub rating: FeedbackRating,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub categories: Option<Vec<FeedbackCategory>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub comment: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct ResponseMetrics {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub capture_ms: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ocr_ms: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub accessibility_ms: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub context_assembly_ms: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub time_to_first_token_ms: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub total_ms: Option<u64>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub input_tokens: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub output_tokens: Option<u32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub context_tokens: Option<u32>,
}

/// Mirrors `BlueyResponse`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct BlueyResponse {
    pub id: String,
    pub request_id: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub session_id: Option<String>,
    pub mode_id: String,
    #[serde(rename = "type")]
    pub response_type: ResponseType,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub title: Option<String>,
    pub content: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub code: Option<CodeBlock>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub sections: Option<Vec<ResponseSection>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub citations: Option<Vec<Citation>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub confidence: Option<f32>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub prompt: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub diagram: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub metrics: Option<ResponseMetrics>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub feedback: Option<ResponseFeedback>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub prepared: Option<bool>,
    /// The output budget cut the answer short (after the engine's one retry
    /// with double the room); `content` is what could be salvaged.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub truncated: Option<bool>,
    pub created_at: String,
}
